File: obs_mcp/scenes.py
#!/usr/bin/env python3

from typing import Any, Dict, List, Optional

from .client import obs_client
from .server import mcp

@mcp.tool()
async def get_scene_list() -> Dict[str, Any]:
    """
    Gets an array of all scenes in OBS.
    
    Returns:
        Dict containing:
        - currentProgramSceneName: Name of the current program scene
        - currentPreviewSceneName: Name of the current preview scene (if studio mode is enabled)
        - scenes: Array of scenes (each with name, sceneIndex)
    """
    return await obs_client.send_request("GetSceneList")

@mcp.tool()
async def get_group_list() -> List[str]:
    """
    Gets an array of all groups in OBS.
    
    Returns:
        List of group names
    """
    response = await obs_client.send_request("GetGroupList")
    return response.get("groups", [])

@mcp.tool()
async def get_current_program_scene() -> str:
    """
    Gets the current program scene.
    
    Returns:
        Name of the current program scene
    """
    response = await obs_client.send_request("GetCurrentProgramScene")
    return response.get("currentProgramSceneName", "")

@mcp.tool()
async def set_current_program_scene(scene_name: str) -> None:
    """
    Sets the current program scene.
    
    Args:
        scene_name: Name of the scene to set as program
    """
    await obs_client.send_request("SetCurrentProgramScene", {"sceneName": scene_name})

@mcp.tool()
async def get_current_preview_scene() -> str:
    """
    Gets the current preview scene (only available when studio mode is enabled).
    
    Returns:
        Name of the current preview scene
    """
    response = await obs_client.send_request("GetCurrentPreviewScene")
    return response.get("currentPreviewSceneName", "")

@mcp.tool()
async def set_current_preview_scene(scene_name: str) -> None:
    """
    Sets the current preview scene (only available when studio mode is enabled).
    
    Args:
        scene_name: Name of the scene to set as preview
    """
    await obs_client.send_request("SetCurrentPreviewScene", {"sceneName": scene_name})

@mcp.tool()
async def create_scene(scene_name: str) -> None:
    """
    Creates a new scene in OBS.
    
    Args:
        scene_name: Name of the scene to create
    """
    await obs_client.send_request("CreateScene", {"sceneName": scene_name})

@mcp.tool()
async def remove_scene(scene_name: str) -> None:
    """
    Removes a scene from OBS.
    
    Args:
        scene_name: Name of the scene to remove
    """
    await obs_client.send_request("RemoveScene", {"sceneName": scene_name})

@mcp.tool()
async def set_scene_name(scene_name: str, new_scene_name: str) -> None:
    """
    Sets the name of a scene (rename).
    
    Args:
        scene_name: Current name of the scene
        new_scene_name: New name for the scene
    """
    await obs_client.send_request("SetSceneName", {
        "sceneName": scene_name,
        "newSceneName": new_scene_name
    })

@mcp.tool()
async def get_scene_scene_transition_override(scene_name: str) -> Dict[str, Any]:
    """
    Gets the scene transition override for a scene.
    
    Args:
        scene_name: Name of the scene
    
    Returns:
        Dict containing transition name and duration (if override exists)
    """
    return await obs_client.send_request("GetSceneSceneTransitionOverride", {"sceneName": scene_name})

@mcp.tool()
async def set_scene_scene_transition_override(scene_name: str, transition_name: Optional[str] = None, 
                                             transition_duration: Optional[int] = None) -> None:
    """
    Sets the scene transition override for a scene.
    
    Args:
        scene_name: Name of the scene
        transition_name: Name of the transition to use, or null to remove
        transition_duration: Duration in milliseconds of the transition, or null to use default
    """
    payload = {"sceneName": scene_name}
    
    if transition_name is not None:
        payload["transitionName"] = transition_name
    
    if transition_duration is not None:
        payload["transitionDuration"] = transition_duration
    
    await obs_client.send_request("SetSceneSceneTransitionOverride", payload)

File: src/tools/scenes.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { OBSWebSocketClient } from "../client.js";
import { z } from "zod";

export async function initialize(server: McpServer, client: OBSWebSocketClient): Promise<void> {
  // GetSceneList tool
  server.tool(
    "obs-get-scene-list",
    "Get a list of scenes in OBS",
    {},
    async () => {
      try {
        const sceneList = await client.sendRequest("GetSceneList");
        return {
          content: [
            {
              type: "text",
              text: JSON.stringify(sceneList, null, 2)
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting scene list: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // GetCurrentProgramScene tool
  server.tool(
    "obs-get-current-scene",
    "Get the current active scene in OBS",
    {},
    async () => {
      try {
        const currentScene = await client.sendRequest("GetCurrentProgramScene");
        return {
          content: [
            {
              type: "text",
              text: `Current scene: ${currentScene.currentProgramSceneName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting current scene: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // SetCurrentProgramScene tool
  server.tool(
    "obs-set-current-scene",
    "Set the current active scene in OBS",
    {
      sceneName: z.string().describe("The name of the scene to set as current")
    },
    async ({ sceneName }) => {
      try {
        await client.sendRequest("SetCurrentProgramScene", { sceneName });
        return {
          content: [
            {
              type: "text",
              text: `Successfully switched to scene: ${sceneName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error setting current scene: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // GetCurrentPreviewScene tool (Studio Mode)
  server.tool(
    "obs-get-preview-scene",
    "Get the current preview scene in OBS Studio Mode",
    {},
    async () => {
      try {
        const previewScene = await client.sendRequest("GetCurrentPreviewScene");
        return {
          content: [
            {
              type: "text",
              text: `Preview scene: ${previewScene.currentPreviewSceneName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error getting preview scene: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // SetCurrentPreviewScene tool (Studio Mode)
  server.tool(
    "obs-set-preview-scene",
    "Set the current preview scene in OBS Studio Mode",
    {
      sceneName: z.string().describe("The name of the scene to set as preview")
    },
    async ({ sceneName }) => {
      try {
        await client.sendRequest("SetCurrentPreviewScene", { sceneName });
        return {
          content: [
            {
              type: "text",
              text: `Successfully set preview scene to: ${sceneName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error setting preview scene: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // CreateScene tool
  server.tool(
    "obs-create-scene",
    "Create a new scene in OBS",
    {
      sceneName: z.string().describe("The name for the new scene")
    },
    async ({ sceneName }) => {
      try {
        await client.sendRequest("CreateScene", { sceneName });
        return {
          content: [
            {
              type: "text",
              text: `Successfully created scene: ${sceneName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error creating scene: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // RemoveScene tool
  server.tool(
    "obs-remove-scene",
    "Remove a scene from OBS",
    {
      sceneName: z.string().describe("The name of the scene to remove")
    },
    async ({ sceneName }) => {
      try {
        await client.sendRequest("RemoveScene", { sceneName });
        return {
          content: [
            {
              type: "text",
              text: `Successfully removed scene: ${sceneName}`
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error removing scene: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );

  // TriggerStudioModeTransition tool
  server.tool(
    "obs-trigger-studio-transition",
    "Trigger a transition from preview to program scene in Studio Mode",
    {},
    async () => {
      try {
        await client.sendRequest("TriggerStudioModeTransition");
        return {
          content: [
            {
              type: "text",
              text: "Successfully triggered studio mode transition"
            }
          ]
        };
      } catch (error) {
        return {
          content: [
            {
              type: "text",
              text: `Error triggering studio transition: ${error instanceof Error ? error.message : String(error)}`
            }
          ],
          isError: true
        };
      }
    }
  );
}